File: content/problems/0103-binary-tree-zigzag-level-order-traversal/solutions/bfs-index-placement/solution.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) return {}; vector<vector<int>> answer; queue<TreeNode*> nodes; nodes.push(root); bool leftToRight = true;
        while (!nodes.empty()) { int size = nodes.size(); vector<int> level(size); for (int index = 0; index < size; ++index) { TreeNode* node = nodes.front(); nodes.pop(); level[leftToRight ? index : size - 1 - index] = node->val; if (node->left) nodes.push(node->left); if (node->right) nodes.push(node->right); } answer.push_back(level); leftToRight = !leftToRight; } return answer;
    }
};

File: content/problems/0103-binary-tree-zigzag-level-order-traversal/solutions/bfs-index-placement/Solution.kt
class Solution {
    fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> {
        if (root == null) return emptyList(); val answer = mutableListOf<List<Int>>(); val queue = ArrayDeque<TreeNode>(); queue.add(root); var leftToRight = true
        while (queue.isNotEmpty()) { val size = queue.size; val level = MutableList(size) { 0 }; repeat(size) { index -> val node = queue.removeFirst(); level[if (leftToRight) index else size - 1 - index] = node.`val`; node.left?.let { queue.add(it) }; node.right?.let { queue.add(it) } }; answer.add(level); leftToRight = !leftToRight }
        return answer
    }
}
